File: frontend/lib/portal-access.ts
/** Staff roles allowed to use the agent portal (website customers are excluded). */
export const PORTAL_STAFF_ROLES = [
  'Booking Agent',
  'Check-in Agent',
  'Support Agent',
  'Pricing Manager',
  'Baggage Handler',
  'Crew Member',
  'System Manager',
  'Administrator',
] as const;

/** Roles that bypass portal doctype permission checks (same as Frappe Desk). */
export const FULL_PORTAL_ACCESS_ROLES = ['System Manager', 'Administrator'] as const;

export function hasPortalAccess(roles: string[] | undefined | null): boolean {
  if (!roles?.length) return false;
  return roles.some((role) =>
    (PORTAL_STAFF_ROLES as readonly string[]).includes(role),
  );
}

export function hasFullPortalPermissions(
  roles: string[] | undefined | null,
  userName?: string | null,
): boolean {
  if (userName === 'Administrator') return true;
  if (!roles?.length) return false;
  return roles.some((role) =>
    (FULL_PORTAL_ACCESS_ROLES as readonly string[]).includes(role),
  );
}

File: bilan_sky/bilan_air_booking_system/utils/portal_access.py
# Copyright (c) 2026, NF and contributors

import frappe
from frappe import _

PORTAL_STAFF_ROLES = frozenset(
	{
		"Booking Agent",
		"Check-in Agent",
		"Support Agent",
		"Pricing Manager",
		"Baggage Handler",
		"Crew Member",
		"System Manager",
		"Administrator",
	}
)

FULL_PORTAL_ACCESS_ROLES = frozenset({"System Manager", "Administrator"})


def user_has_full_portal_permissions(user=None):
	"""Administrator and System Manager bypass portal doctype permission checks."""
	user = user or frappe.session.user
	if not user or user == "Guest":
		return False
	if user == "Administrator":
		return True
	return bool(FULL_PORTAL_ACCESS_ROLES.intersection(frappe.get_roles(user)))


def user_has_portal_access(user=None):
	user = user or frappe.session.user
	if not user or user == "Guest":
		return False
	return bool(PORTAL_STAFF_ROLES.intersection(frappe.get_roles(user)))


def ensure_portal_system_user(user=None):
	"""Portal staff must be System User for Role Permission Manager rules to apply."""
	user = user or frappe.session.user
	if not user_has_portal_access(user):
		return
	if frappe.db.get_value("User", user, "user_type") == "System User":
		return
	frappe.db.set_value("User", user, "user_type", "System User", update_modified=False)
	frappe.clear_cache(user=user)
	if getattr(frappe.local, "role_permissions", None) is not None:
		frappe.local.role_permissions = {}


def require_portal_staff():
	if frappe.session.user == "Guest":
		frappe.throw(_("Please sign in to the agent portal."), frappe.PermissionError)
	if not user_has_portal_access():
		frappe.throw(
			_("The agent portal is restricted to booking agents and staff."),
			frappe.PermissionError,
		)
